File: frontend/js/views/cap_table.js
// Cap table generator — per-holder ownership % of fully-diluted shares, via
// /calc/cap-table. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const num = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
let LAST_DOC = null;

const CLASSES = ['Common', 'Preferred', 'Options'];
const SEED = [
    { name: 'Founder', shares: 800000, cls: 'Common' },
    { name: 'Investor', shares: 200000, cls: 'Preferred' },
];

function classOpts(sel) {
    return CLASSES.map((c) => `<option value="${c}" ${c === sel ? 'selected' : ''}>${c}</option>`).join('');
}

function rowHtml(h) {
    return `
        <div class="mpb-row ick-row">
            <input type="text" class="ct-name" placeholder="${esc(t('view.ct.ph.name'))}" value="${esc(h.name || '')}">
            <input type="number" step="1" min="0" class="ct-shares" value="${h.shares}">
            <select class="ct-class">${classOpts(h.cls || 'Common')}</select>
            <button type="button" class="ct-del" data-i18n="view.ct.remove">Remove</button>
        </div>`;
}

export async function renderCapTable(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.ct.h1.title">// CAP TABLE</span></h1>
        <p class="muted small" data-i18n="view.ct.hint.intro">
            The ledger of who owns a company's equity. Each holder's shares as a percentage of the
            fully-diluted total (issued shares plus the unallocated option pool) gives their ownership.
            Drafting aid, not legal/securities advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.ct.h2.inputs">Capitalization</h2>
            <form id="ct-form" class="inline-form">
                <label><span data-i18n="view.ct.label.company">Company</span>
                    <input type="text" name="company_name" value="Widgets Inc" required></label>
                <label><span data-i18n="view.ct.label.date">As of date</span>
                    <input type="date" name="as_of_date" value="2026-07-15" required></label>
                <label><span data-i18n="view.ct.label.pool">Option pool (unallocated shares)</span>
                    <input type="number" step="1" min="0" name="option_pool_shares" value="100000"></label>
            </form>
            <div class="mpb-head ick-head">
                <span data-i18n="view.ct.col.name">Holder</span>
                <span data-i18n="view.ct.col.shares">Shares</span>
                <span data-i18n="view.ct.col.class">Class</span>
                <span></span>
            </div>
            <div id="ct-rows">${SEED.map(rowHtml).join('')}</div>
            <button type="button" id="ct-add" class="secondary" data-i18n="view.ct.add">+ Add holder</button>
        </div>
        <div id="ct-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#ct-form');
    const rowsEl = mount.querySelector('#ct-rows');

    const generate = async () => {
        const fd = new FormData(form);
        const holders = [...rowsEl.querySelectorAll('.ick-row')].map((r) => ({
            name: (r.querySelector('.ct-name').value || '').trim(),
            shares: Number(r.querySelector('.ct-shares').value) || 0,
            class: r.querySelector('.ct-class').value || 'Common',
        })).filter((h) => h.name);
        const body = {
            company_name: (fd.get('company_name') || '').trim(),
            as_of_date: fd.get('as_of_date'),
            holders,
            option_pool_shares: Number(fd.get('option_pool_shares')) || 0,
        };
        try {
            const doc = await api.calcCapTable(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.ct.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);

    mount.querySelector('#ct-add').addEventListener('click', () => {
        rowsEl.insertAdjacentHTML('beforeend', rowHtml({ name: '', shares: 0, cls: 'Common' }));
        applyUiI18n(rowsEl.lastElementChild);
        generate();
    });
    rowsEl.addEventListener('click', (e) => {
        if (e.target.classList.contains('ct-del')) {
            e.target.closest('.ick-row').remove();
            generate();
        }
    });
    form.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('change', () => { live(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#ct-result');
    const rows = doc.holders.map((h) => `
        <tr><td>${esc(h.name)}</td><td>${esc(h.class)}</td><td>${num(h.shares)}</td><td>${pct(h.ownership_pct)}</td></tr>
    `).join('');
    const poolRow = doc.option_pool_shares > 0
        ? `<tr><td data-i18n="view.ct.pool">Option pool</td><td>—</td><td>${num(doc.option_pool_shares)}</td><td>${pct(doc.option_pool_pct)}</td></tr>`
        : '';
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.ct.card.fd">Fully-diluted</div>
                    <div class="value">${num(doc.total_fully_diluted_shares)}</div></div>
                <div class="card"><div class="label" data-i18n="view.ct.card.issued">Issued</div>
                    <div class="value">${num(doc.total_issued_shares)}</div></div>
                <div class="card"><div class="label" data-i18n="view.ct.card.holders">Holders</div>
                    <div class="value">${doc.holder_count}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="ct-copy" type="button" data-i18n="view.ct.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="ct-download" type="button" data-i18n="view.ct.btn.download">Download .txt</button>
            </div>
        </div>
        <table class="data-table">
            <thead><tr>
                <th data-i18n="view.ct.th.name">Holder</th>
                <th data-i18n="view.ct.th.class">Class</th>
                <th data-i18n="view.ct.th.shares">Shares</th>
                <th data-i18n="view.ct.th.own">Ownership</th>
            </tr></thead>
            <tbody>${rows}${poolRow}</tbody>
        </table>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#ct-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.ct.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.ct.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#ct-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'cap-table.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/cap_table.rs
//! Capitalization table (cap table) — the ledger of who owns a company's
//! equity. Each holder's shares as a percentage of the fully-diluted total
//! (issued shares plus the unallocated option pool) gives their ownership.
//! Distinct from the subscription/operating agreements (this is the resulting
//! ownership ledger). Pure compute plus a document layout. Drafting aid, not
//! legal/securities advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Holder {
    pub name: String,
    pub shares: f64,
    /// Security class label (Common, Preferred, Options, …).
    #[serde(default)]
    pub class: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CapTableInput {
    pub company_name: String,
    pub as_of_date: String,
    #[serde(default)]
    pub holders: Vec<Holder>,
    /// Unallocated option pool shares (counted in the fully-diluted total).
    #[serde(default)]
    pub option_pool_shares: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct HolderRow {
    pub name: String,
    pub shares: f64,
    pub class: String,
    pub ownership_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CapTable {
    pub title: String,
    pub total_issued_shares: f64,
    pub option_pool_shares: f64,
    pub total_fully_diluted_shares: f64,
    pub holder_count: usize,
    pub holders: Vec<HolderRow>,
    pub option_pool_pct: f64,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn commas(n: f64) -> String {
    let n = n.round() as i64;
    let s = n.abs().to_string();
    let mut out = String::new();
    for (idx, ch) in s.chars().enumerate() {
        if idx > 0 && (s.len() - idx).is_multiple_of(3) {
            out.push(',');
        }
        out.push(ch);
    }
    if n < 0 {
        format!("-{out}")
    } else {
        out
    }
}

pub fn generate(i: &CapTableInput) -> CapTable {
    let total_issued: f64 = i.holders.iter().map(|h| h.shares).sum();
    let fully_diluted = total_issued + i.option_pool_shares;

    let pct = |shares: f64| -> f64 {
        if fully_diluted > 0.0 {
            cents(shares / fully_diluted * 100.0)
        } else {
            0.0
        }
    };

    let holders: Vec<HolderRow> = i
        .holders
        .iter()
        .map(|h| HolderRow {
            name: h.name.clone(),
            shares: h.shares,
            class: if h.class.trim().is_empty() { "Common".to_string() } else { h.class.trim().to_string() },
            ownership_pct: pct(h.shares),
        })
        .collect();

    let option_pool_pct = pct(i.option_pool_shares);

    let table_body = if holders.is_empty() {
        "No holders recorded.".to_string()
    } else {
        let mut lines: Vec<String> = holders
            .iter()
            .map(|h| format!("  • {} ({}): {} shares — {:.2}%", h.name, h.class, commas(h.shares), h.ownership_pct))
            .collect();
        if i.option_pool_shares > 0.0 {
            lines.push(format!("  • Option pool (unallocated): {} shares — {:.2}%", commas(i.option_pool_shares), option_pool_pct));
        }
        lines.join("\n")
    };

    let clauses = vec![
        DocClause {
            heading: "Header".into(),
            body: format!("Company: {}\nCapitalization as of: {}", i.company_name, i.as_of_date),
        },
        DocClause { heading: "1. Capitalization".into(), body: table_body },
        DocClause {
            heading: "2. Summary".into(),
            body: format!(
                "Issued shares: {}\nOption pool (unallocated): {}\nFully-diluted total: {}",
                commas(total_issued), commas(i.option_pool_shares), commas(fully_diluted)
            ),
        },
        DocClause {
            heading: "3. Certification".into(),
            body: "This table reflects the Company's capitalization as of the date stated, to the best of the Company's knowledge. It is for reference and is not a substitute for the Company's official stock records.".into(),
        },
        DocClause {
            heading: "Signature".into(),
            body: format!("Prepared by: ____________________  Date: {}\n{}", i.as_of_date, i.company_name),
        },
    ];

    CapTable {
        title: "Capitalization Table".into(),
        total_issued_shares: total_issued,
        option_pool_shares: i.option_pool_shares,
        total_fully_diluted_shares: fully_diluted,
        holder_count: holders.len(),
        holders,
        option_pool_pct,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn holder(name: &str, shares: f64, class: &str) -> Holder {
        Holder { name: name.into(), shares, class: class.into() }
    }

    fn base() -> CapTableInput {
        CapTableInput {
            company_name: "Widgets Inc".into(),
            as_of_date: "2026-07-15".into(),
            holders: vec![
                holder("Founder", 800_000.0, "Common"),
                holder("Investor", 200_000.0, "Preferred"),
            ],
            option_pool_shares: 100_000.0,
        }
    }

    #[test]
    fn totals_and_ownership() {
        let d = generate(&base());
        assert!(close(d.total_issued_shares, 1_000_000.0));
        assert!(close(d.total_fully_diluted_shares, 1_100_000.0));
        assert!(close(d.holders[0].ownership_pct, 72.73));
        assert!(close(d.holders[1].ownership_pct, 18.18));
        assert!(close(d.option_pool_pct, 9.09));
    }

    #[test]
    fn percentages_sum_to_100() {
        let d = generate(&base());
        let sum: f64 = d.holders.iter().map(|h| h.ownership_pct).sum::<f64>() + d.option_pool_pct;
        assert!(close(sum, 100.0));
    }

    #[test]
    fn table_lists_holders_and_pool() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading == "1. Capitalization").unwrap();
        assert!(c.body.contains("Founder (Common): 800,000 shares — 72.73%"));
        assert!(c.body.contains("Option pool (unallocated): 100,000 shares — 9.09%"));
    }

    #[test]
    fn no_pool_omits_pool_line() {
        let d = generate(&CapTableInput { option_pool_shares: 0.0, ..base() });
        // Without a pool, fully-diluted = issued; founder is 80%.
        assert!(close(d.holders[0].ownership_pct, 80.0));
        let c = d.clauses.iter().find(|c| c.heading == "1. Capitalization").unwrap();
        assert!(!c.body.contains("Option pool"));
    }

    #[test]
    fn default_class_is_common() {
        let d = generate(&CapTableInput {
            holders: vec![holder("X", 100.0, "")],
            option_pool_shares: 0.0,
            ..base()
        });
        assert_eq!(d.holders[0].class, "Common");
    }

    #[test]
    fn empty_holders() {
        let d = generate(&CapTableInput { holders: vec![], option_pool_shares: 0.0, ..base() });
        assert_eq!(d.holder_count, 0);
        let c = d.clauses.iter().find(|c| c.heading == "1. Capitalization").unwrap();
        assert!(c.body.contains("No holders"));
    }
}
